Implement a functional filesystem for the cFE

base0x10 opened the pull request:
### Summary of this PR

The cFE requires that the osal provide a nearly posix compliant filesystem.  This includes abstractions for directory streams, file descriptors, and physical volumes.  I have provided a filesystem implementation which fulfils these abstractions and implements all of the required API.  

There is more work to be done on this code.  It is of high quality, but I am welcome to criticism and corrections about style.  Wherever appropriate, I have used cFE and OSAL naming conventions and comment conventions above composite ones.  A more general implementation of filedes is required for comparability as well as to allow concurrent access to files.  

I would appreciate feedback, @others, @zacharied and @gparmer 

### Code Quality

As part of this pull request, I've considered the following:

Style:

- [ x] Comments adhere to the Style Guide (SG)
- [ x] Spacing adhere's to the SG 
- [x ] Naming adhere's to the SG
- [ x] All other aspects of the SG are adhered to, or exceptions are justified in this pull request

Code Craftsmanship:

- [ x] I've made an attempt to remove all redundant code
- [ x] I've considered ways in which my changes might impact existing code, and cleaned it up
- [ x] I've formatted the code in an effort to make it easier to read (proper error handling, function use, etc...)
- [ x] I've commented appropriately where code is tricky
- [ x] I agree that there is no "throw-away" code, and that code in this PR is of high quality

base0x10 committed:
Implement a functional filesystem for the cFE

The cFE requires that the osal provide a nearly posix compliant
filesystem.  This includes abstractions for directory streams,
file descriptors, and physical volumes.  I have provided a
filesystem implementation which fulfills these abstractions and
implements all of the required API.

Others commented on src/components/implementation/no_interface/cFE_booter/osapi.c:
```diff
@@ -25,6 +25,7 @@ int32 OS_API_Init(void)
 
         cos_meminfo_init(&(ci->mi), BOOT_MEM_KM_BASE, COS_MEM_KERN_PA_SZ, BOOT_CAPTBL_SELF_UNTYPED_PT);
         cos_defcompinfo_init();
+        OS_FS_Init();
```
I'd put spaces around this new line.

Others commented on src/components/implementation/no_interface/cFE_booter/osfiles.c:
```diff
@@ -17,7 +22,9 @@ int32 OS_FS_Init(void)
 {
-    PANIC("Read Only Filesystem");
+    int32 ret = chk_path_new(path);
+    if (ret) return ret;
```
Check explicitly against success. Also, I don't know if you really need a variable here...

base0x10 replied:
Fixed checking against success but there are three error values that could be returned.  I need the variable

Others commented on src/components/implementation/no_interface/cFE_booter/osfiles.c:
```diff
@@ -26,70 +33,71 @@ int32 OS_creat(const char *path, int32  access)
 {
-    PANIC("Read Only Filesystem!");
-    return 0;
+    return file_write(filedes, buffer, nbytes);
```
I'm not sure how I feel about all these methods that are basically wrappers...

base0x10 replied:
I would like to talk more about what a sensible way to refactor it into modularized but non-redundant code

zacharied replied:
yeah, why exactly are we just wrapping methods that you're defining anyways? I may just be missing something

gparmer replied:
Note that this will be essential once the fs stuff is a non-cFE-specific library.  So I don't know if the question is if we should have wrapping, rather how to do it best

Others commented on src/components/implementation/no_interface/cFE_booter/osfiles.c:
```diff
@@ -26,70 +33,71 @@ int32 OS_creat(const char *path, int32  access)
-    return 0;
+    if (filestats == NULL) return OS_FS_ERR_INVALID_POINTER;
+    int32 ret = chk_path(path);
+    if (ret) return ret;
```
Check explicitly against success.

Others commented on src/components/implementation/no_interface/cFE_booter/osfiles.c:
```diff
@@ -98,120 +106,160 @@ int32 OS_remove(const char *path)
+    int32 ret = chk_path(old_filename);
+    if (ret) return ret;
+    ret = chk_path(new_filename);
+    if (ret) return ret;
```
Check explicitly against success

Others commented on src/components/implementation/no_interface/cFE_booter/osfiles.c:
```diff
@@ -98,120 +106,160 @@ int32 OS_remove(const char *path)
+    if (strlen((char *)src) > OS_MAX_PATH_LEN) return OS_FS_ERR_PATH_TOO_LONG;
+    if (strlen((char *)dest) > OS_MAX_PATH_LEN) return OS_FS_ERR_PATH_TOO_LONG;
+    if (strlen(path_to_name((char *)src)) > OS_MAX_FILE_NAME) return OS_FS_ERR_NAME_TOO_LONG;
+    if (strlen(path_to_name((char *)dest)) > OS_MAX_FILE_NAME) return OS_FS_ERR_NAME_TOO_LONG;
```
You shouldn't have to cast away const so much. Can you change the signature of the functions you're calling? Or cast once at the top of the function?

Others commented on src/components/implementation/no_interface/cFE_booter/osfiles.c:
```diff
@@ -98,120 +106,160 @@ int32 OS_remove(const char *path)
+    int32 ret = chk_path(src);
+    if (ret) return ret;
+    ret = chk_path_new(dest);
+    if (ret) return ret;
```
Check explicitly against success, that's cFE style.

Others commented on src/components/implementation/no_interface/cFE_booter/osfiles.c:
```diff
@@ -98,120 +106,160 @@ int32 OS_remove(const char *path)
+    if (strlen((char *)src) > OS_MAX_PATH_LEN) return OS_FS_ERR_PATH_TOO_LONG;
+    if (strlen((char *)dest) > OS_MAX_PATH_LEN) return OS_FS_ERR_PATH_TOO_LONG;
+    if (strlen(path_to_name((char *)src)) > OS_MAX_FILE_NAME) return OS_FS_ERR_NAME_TOO_LONG;
+    if (strlen(path_to_name((char *)dest)) > OS_MAX_FILE_NAME) return OS_FS_ERR_NAME_TOO_LONG;
```
Same problem as above. Also, I think you might be able to factor this out to another function.

base0x10 replied:
The pattern is only used twice, so I don't think it would be more useful to factor out.  In every other place it is factored out to chkpath, but the unit tests care about the order of the checks, so I cannot check for all bad inputs on src, and then check for bad inputs on dest

Others commented on src/components/implementation/no_interface/cFE_booter/osfiles.c:
```diff
@@ -98,120 +106,160 @@ int32 OS_remove(const char *path)
-    PANIC("Unimplemented method!"); // TODO: Implement me!
-    return 0;
+    int32 ret = chk_path(Filename);
+    if (ret) return OS_FS_ERR_INVALID_POINTER;
```
This pattern is ridiculously common in your code. Could we factor this out to a macro? Is that even a good idea?

base0x10 replied:
That can be up for discussion.  I don't like macros, but maybe it is the right choice.   In the refactor we discussed, nearly all of this pattern will be eliminated, so it probably wont be a problem.

Others commented on src/components/implementation/no_interface/cFE_booter/osfiles.c:
```diff
@@ -98,120 +106,160 @@ int32 OS_remove(const char *path)
+    if (ret) return OS_FS_ERR_INVALID_POINTER;
+
+    struct fsobj *file = file_find(Filename);
+    if (file == NULL) return OS_INVALID_POINTER;
```
It's a little inconsistent to compare to null explicitly here, and use `!ptr` in other places...

base0x10 replied:
valid

Others commented on src/components/implementation/no_interface/cFE_booter/osfiles.c:
```diff
@@ -98,120 +106,160 @@ int32 OS_remove(const char *path)
-    PANIC("Unimplemented method!"); // TODO: Implement me!
-    return 0;
+    if (directory == NULL) return OS_FS_ERR_INVALID_POINTER;
+    return file_close((int32) directory);
```
Are you casting a pointer into an integer value here???

Others commented on src/components/implementation/no_interface/cFE_booter/osfiles.c:
```diff
@@ -98,120 +106,160 @@ int32 OS_remove(const char *path)
 void OS_rewinddir(os_dirp_t directory)
 {
-    PANIC("Unimplemented method!"); // TODO: Implement me!
+    file_rewinddir();
```
Why do we ignore the argument here? I'd add a comment explaining that.

Others commented on src/components/implementation/no_interface/cFE_booter/osfiles.c:
```diff
@@ -98,120 +106,160 @@ int32 OS_remove(const char *path)
-    PANIC("Unimplemented method!"); // TODO: Implement me!
-    return 0;
+    if (directory == NULL) return NULL;
+    return file_readdir( (int32) directory);
```
Same questions about casting here

Others commented on src/components/implementation/no_interface/cFE_booter/osfiles.c:
```diff
@@ -270,16 +313,19 @@ int32 OS_rmfs(char *devname)
+    int32 ret = chk_path(name);
+    if (ret) return ret;
+
+    return 10;
```
This feels sketchy AF. Even if for some reason randomly returning 10 is always the right thing to do, this should be a defined constant, not a magic value.

Others commented on src/components/implementation/no_interface/cFE_booter/osfiles.c:
```diff
@@ -270,16 +313,19 @@ int32 OS_rmfs(char *devname)
-    return 0;
+
+    int32 ret = chk_path(name);
+    if (ret) return ret;
```
Last comment about this, but you should always compare explicitly against success. That's cFE style!

Others commented on src/components/implementation/no_interface/cFE_booter/osfiles.c:
```diff
@@ -289,8 +335,11 @@ int32 OS_fsBlocksFree(const char *name)
+    int32 ret = chk_path(name);
+    if (ret) return ret;
+
+    return 10 * F_PART_DATA_SIZE;
```
I think you should base this on the `fsBlocksFree` function

Others commented on src/components/implementation/no_interface/cFE_booter/osfiles.c:
```diff
@@ -289,8 +335,11 @@ int32 OS_fsBlocksFree(const char *name)
 {
-    PANIC("Unimplemented method!"); // TODO: Implement me!
-    return 0;
+    if (bytes_free == NULL) return OS_FS_ERR_INVALID_POINTER;
```
More inconsistency with the null checks. You should pick one or the other. (Look at how osapi.c and ostask.c do it?) Last comment on this, but it should be fixed throughout.

Others commented on src/components/implementation/no_interface/cFE_booter/osfiles.c:
```diff
@@ -299,17 +348,18 @@ int32 OS_fsBytesFree(const char *name, uint64 *bytes_free)
 {
-    PANIC("Unimplemented method!"); // TODO: Implement me!
-    return 0;
+    return OS_ERR_NOT_IMPLEMENTED;
```
Can you not just write a dummy implementation of this that always says the filesystem is healthy?

base0x10 replied:
Posix osal does the same thing, so I think it would be more consistent to not change it

Others commented on src/components/implementation/no_interface/cFE_booter/osfiles.c:
```diff
@@ -327,8 +381,8 @@ int32 OS_TranslatePath(const char *VirtualPath, char *LocalPath)
-    PANIC("Unimplemented method!"); // TODO: Implement me!
-    return 0;
+    if (filesys_info == NULL) return OS_FS_ERR_INVALID_POINTER;
+    return Filesys_GetFsInfo(filesys_info);
```
So many wrapping methods, holy cow. These `Filesys_` ones feel especially unnecessary.

base0x10 replied:
agreed, agreed, and we will be refactoring some

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.h:
```diff
@@ -1,9 +1,15 @@
+
+#include <cos_component.h>
+#include <cos_defkernel_api.h>
+#include <cos_kernel_api.h>
```
Your includes are in the wrong order. It should be:

1. libc
2. gen
3. our headers

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.h:
```diff
@@ -12,6 +18,9 @@
 #define INT32_MAX 0x7FFFFFF //2^31 - 1
 
+// a page is 4096, size of f_part is 5 values * 4 bytes
+#define F_PART_DATA_SIZE 4096-4*5
```
Shouldn't this be something like `FILESYS_BLOCK_SIZE` since that's the terminology used elsewhere? Also this macro is egregiously unhygienic... Remember parenthesis!

base0x10 replied:
It seems to me that most of the places that this is used f_part_datasize more clearly describes what is going on.  internally when it shows up its important that the reader can see it is referring to the len of the array within f_part.  also, does blocksize usually refer to the size of the blocks including the struct data or not?  Tar includes metadata in blocksize, but not every block starts with a header.  I am open to discussion.

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.h:
```diff
@@ -27,41 +36,64 @@ typedef enum {
+    struct file_position position;
+    uint32 refcnt; //number of threads which have opened it
+    uint32 mode;
+    struct f_part *file_part;
```
I find this name confusing... What does it mean?

base0x10 replied:
it is a part of a file.  added a comment

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.h:
```diff
@@ -27,41 +36,64 @@ typedef enum {
+typedef enum {
+    STATIC,
+    DYNAMIC,
+} fpart_type;
```
Nice use of an enum

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.h:
```diff
@@ -27,41 +36,64 @@ typedef enum {
+    struct f_part *file_part;
     struct fsobj *next, *prev;
-    struct fsobj *child, *parent;   /* child != NULL iff type = dir */
+    struct fsobj *child, *parent; // child != NULL iff type = dir
```
What do all these fsobj variables mean?

base0x10 replied:
the filesystem is a linked list.  but I added some comments

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.h:
```diff
@@ -27,41 +36,64 @@ typedef enum {
+
+struct f_part {
+    struct fsobj *file;
+    struct f_part *next, *prev;
```
This is like the perfect usecase for `ps_list`. I know it's probably a big change, but it'd cut down on code significantly. (And probably remove bugs too.)

base0x10 replied:
Yeah, that is on the queue of eventual changes

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.h:
```diff
@@ -27,41 +36,64 @@ typedef enum {
 
+uint32 FS_entrypoint();
+uint32 FS_posttest();
+uint32 BFT(struct fsobj *o);
```
What does this acronym mean?

base0x10 replied:
My bad.  BFT was breadth first traversal but it is not used.

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.h:
```diff
@@ -27,41 +36,64 @@ typedef enum {
 
 int32 tar_read(uint32 offset, char *buf, uint32 req_sz);
 
 uint32 tar_cphdr(uint32 tar_offset, struct fsobj *file);
```
I'd use one line between each of these methods.

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.h:
```diff
@@ -80,34 +112,126 @@ static inline uint32 oct_to_dec(char *oct)
+
+static inline char *path_to_name(char *path){
+
+    uint32 path_len = strlen(path), offset;
```
Why would you calculate `strlen(path)` and then throw away the result?

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.h:
```diff
@@ -80,34 +112,126 @@ static inline uint32 oct_to_dec(char *oct)
+static inline char *path_to_name(char *path){
+
+    uint32 path_len = strlen(path), offset;
+    assert(path[path_len - 2] != '/' && path_len > 1);
```
I think you should actually handle this case, not just assert and blow up.

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.h:
```diff
@@ -80,34 +112,126 @@ static inline uint32 oct_to_dec(char *oct)
+        //do nothing
+    }
+    assert(0 < strlen(path + offset + 1));
+    return path + offset + 1;
```
This weird string manipulation needs a lot more comments. (What is this method even doing???) Also it feels like there is definitely a better way of doing this...

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.h:
```diff
@@ -80,34 +112,126 @@ static inline uint32 oct_to_dec(char *oct)
+    assert(o && o->child == NULL);
+
+    // if o is first in list of children
+    if (o->prev == NULL && o->parent) {
```
Why do you check that o->parent is not null here? Wouldn't an assert be better? (From my understanding, parent is only null when o == ROOT)

base0x10 replied:
The reason we are checking is in case they call rm with ROOT.  the first if statement checks if o is the first child in a list of children, and if o has a parent.  If these are true, it will change the parent's 'child' pointer to the second element in the linked list of children.  This step is unnecessary and will cause null pointer errors or logical errors if either o == root or o is not the first child in its list

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.h:
```diff
@@ -80,34 +112,126 @@ static inline uint32 oct_to_dec(char *oct)
+    o->size = 0;
+    o->refcnt = 0;
+    o->mode = 0;
+    o->file_part = NULL;
```
Do we really need to take the time to reset all these fields? If we reuse the fsobj, won't they just be overwritten anyway?

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.h:
```diff
@@ -80,34 +112,126 @@ static inline uint32 oct_to_dec(char *oct)
+
+int32 Filesys_GetPhysDriveName(char * PhysDriveName, char * MountPoint);
+
+int32 Filesys_GetFsInfo(os_fsinfo_t *finesys_info);
```
These two method names are inconsistent with the rest of the api.

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.h:
```diff
@@ -80,34 +112,126 @@ static inline uint32 oct_to_dec(char *oct)
 
 uint32 newfs_init(char *defilvname, char *volname, uint32 blocksize, uint32 numblocks);
+
+int32 rmfs(char *devname);
```
Standardize on method names. I'd recommend all methods start with: `OS_`, `file_` or `filesystem_`. Lot's of methods do not use this convention, and it causes unnecessary confusion.

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.h:
```diff
@@ -80,34 +112,126 @@ static inline uint32 oct_to_dec(char *oct)
+    if (path == NULL) return OS_FS_ERR_INVALID_POINTER;
+    if (strlen(path) > OS_MAX_PATH_LEN) return OS_FS_ERR_PATH_TOO_LONG;
+    return OS_FS_SUCCESS;
+}
```
You have these methods, but don't use them consistently... Either remove them or use them.

base0x10 replied:
I do use these consistently.  Mostly in osfiles.c where I am checking inputs.

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.h:
```diff
@@ -80,34 +112,126 @@ static inline uint32 oct_to_dec(char *oct)
+    o->mode = 0;
+    o->file_part = NULL;
+    return OS_FS_SUCCESS;
+}
```
I have concerns with the number of static inline functions in the header. Many of these seem internal / only used in one file, in which case they can just be defined in the C file they're used in. Additionally some of these aren't even really used or, if they are, aren't used in performance critical bits. We shouldn't overuse static inline unless we need to.

base0x10 replied:
Why whats the drawback to using too many static inline?  Object file size?  But I guess it would make more sense to put these methods inside of osfilesys.c rather than declare in osfilesys.h

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -1,19 +1,11 @@
 #include "osfilesys.h"
 
-struct FS fs[MAX_NUM_FS];
+struct fs filesystem[MAX_NUM_FS];
```
Shouldn't this be `filesystems`?

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -1,19 +1,11 @@
-struct FS fs[MAX_NUM_FS];
+struct fs filesystem[MAX_NUM_FS];
+struct fs *openfs;
+struct fsobj *openfile;
```
Can you only have one file open at a time?

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -42,32 +34,41 @@ uint32 tar_load()
-    if (!fs[0].devname) return OS_FS_ERR_DRIVE_NOT_CREATED;
-
-    struct fsobj *o = fs[0].root;
+    uint32 i;
```
Where do you use this variable?

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -79,52 +80,60 @@ uint32 tar_parse()
  */
 uint32 tar_cphdr(uint32 tar_offset, struct fsobj *file)
```
What's wrong with `tar_copy_header`?

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -79,52 +80,60 @@ uint32 tar_parse()
+        file->size              = oct_to_dec(location + 124);
+        file->file_part         = part;
+        file->file_part->data   = location + 500;
     }
```
Add some comments around the magic values here. (Maybe just one big one describing the tar header.)

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -79,52 +80,60 @@ uint32 tar_parse()
-{
-    if (offset % TAR_BLOCKSIZE) return offset + (TAR_BLOCKSIZE - (offset % TAR_BLOCKSIZE));
-    return offset;
-}
```
I don't understand why this was moved to the header.

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -79,52 +80,60 @@ uint32 tar_parse()
-    o->next     = o->prev = NULL;
+    if (ci == NULL) {
+        struct cos_defcompinfo *defci = cos_defcompinfo_curr_get();
+        ci = &(defci->ci);
```
Parentheses are unnecessary here.

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -79,52 +80,60 @@ uint32 tar_parse()
+    (*part)->file = NULL;
+    (*part)->data = (char *) part + sizeof(part);
     return OS_FS_SUCCESS;
 }
```
This method always returns `OS_FS_SUCCESS`. I think you should return the pointer instead. (This would avoid the need for double pointers.)

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -138,28 +147,25 @@ uint32 newfile_get(struct fsobj **o)
+        openfs->root = o;
+        return OS_FS_SUCCESS;
+    }
+    assert(o && path && openfs);
```
You've already used openfs, what's the point of check if it's null?

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -185,75 +195,483 @@ uint32 insert(struct fsobj *o, struct fsobj *root)
-    root->refcnt++;
-    return root->filedes;
+int32 file_open(const char *path) {
+    assert(openfs);
```
What's the point of checking if a variable is null, if you're going to immediately write to it?

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -185,75 +195,483 @@ uint32 insert(struct fsobj *o, struct fsobj *root)
+int32 file_close(int32 filedes) {
+
+    int32 ret = chk_fd(filedes);
+    if (ret) return ret;
```
Check explicitly against success.

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -185,75 +195,483 @@ uint32 insert(struct fsobj *o, struct fsobj *root)
+{
+    if (!buffer)                                    return OS_FS_ERR_INVALID_POINTER;
+    int32 ret = chk_fd(filedes);
+    if (ret) return ret;
```
Check explicitly against success.

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -185,75 +195,483 @@ uint32 insert(struct fsobj *o, struct fsobj *root)
+
+    PANIC("Unreachable Statement");
+    return 0;
+}
```
This function assumes all the error codes are negative. Is that safe?

base0x10 replied:
yes.  both the documentation and unit tests require OS_FS_ERROR and OS_FS_ERR_INVALID_POINTER.  also hard coded values are negative

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -185,75 +195,483 @@ uint32 insert(struct fsobj *o, struct fsobj *root)
+
+                memcpy(buffer, &part->data[o->position.part_offset], read_size);
+
+                buffer                      += read_size;
```
The spacing is pretty extreme here. Spaces add clarity, but there are too many here. The ones after ifs and whiles feel especially unnecessary.

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -185,75 +195,483 @@ uint32 insert(struct fsobj *o, struct fsobj *root)
+    if (openfile->parent == NULL) return;
+    assert(openfile->parent->child);
+    openfile = openfile->parent->child;
+    return;
```
I don't think you need this trailing return.

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -185,75 +195,483 @@ uint32 insert(struct fsobj *o, struct fsobj *root)
+    if (path[0] != '/') return NULL;
+    path++;
+
+    assert(path);
```
Shouldn't this go above where you use path for the first time?

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -185,75 +195,483 @@ uint32 insert(struct fsobj *o, struct fsobj *root)
             if (!strcmp(root->name, path)) return root;
             if (!root->next) return NULL;
             root = root->next;
         }
```
This would be a lot clearer if you were using ps_list.

base0x10 replied:
I agree that we should switch to ps list.

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -185,75 +195,483 @@ uint32 insert(struct fsobj *o, struct fsobj *root)
 
+        //root must now be parent of final location
+        if (!strcmp(root->name, path)) return root;
+        assert(strspn(path, root->name) >= strlen(root->name));
```
This assert doesn't make sense to me.

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -185,75 +195,483 @@ uint32 insert(struct fsobj *o, struct fsobj *root)
+{
+    assert(path);
+    struct fsobj *o;
+    if (newfile_get(&o)) return OS_FS_ERR_DRIVE_NOT_CREATED;
```
Check explicitly against success.

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -185,75 +195,483 @@ uint32 insert(struct fsobj *o, struct fsobj *root)
+    o->name = path_to_name((char *) path);
+    o->type = FSOBJ_DIR;
+    o->size = 0;
+    if (file_insert(o, (char *) path)) return OS_FS_ERR_PATH_INVALID;
```
Check explicitly against success.

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -185,75 +195,483 @@ uint32 insert(struct fsobj *o, struct fsobj *root)
+    assert(path);
+    struct fsobj *o;
+    if (newfile_get(&o)) return OS_FS_ERR_DRIVE_NOT_CREATED;
+    o->name = path_to_name((char *) path);
```
You're defining both these functions, you should never need to cast away const...

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -185,75 +195,483 @@ uint32 insert(struct fsobj *o, struct fsobj *root)
+            if (cur->prev != NULL) {
+                assert(cur->prev->next == cur);
+                cur = cur->prev;
+                rm(cur->next);
```
`rm` probably should be namespaced

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -185,75 +195,483 @@ uint32 insert(struct fsobj *o, struct fsobj *root)
+    return OS_FS_SUCCESS;
+}
+
+int32 file_creat(const char *path)
```
Just use `create` for internal APIs. Don't perpetuate mistaken naming lol

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -185,75 +195,483 @@ uint32 insert(struct fsobj *o, struct fsobj *root)
+    assert(path);
+    struct fsobj *o;
+    if (newfile_get(&o)) return OS_FS_ERR_DRIVE_NOT_CREATED;
+    o->name = (char *) path;
```
If you have to cast away const, don't take const parameters. This needs to be fixed throughout.

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -185,75 +195,483 @@ uint32 insert(struct fsobj *o, struct fsobj *root)
+{
+    struct fsobj *file = file_find(path);
+    if (!file) return OS_FS_ERR_PATH_INVALID;
+    rm(file);
```
`file_rm`?

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -185,75 +195,483 @@ uint32 insert(struct fsobj *o, struct fsobj *root)
+{
+    struct fsobj *file = file_find(old_filename);
+    if (!file) return OS_FS_ERR_PATH_INVALID;
+    file->name = path_to_name((char *)new_filename);
```
You should never have to cast away const in a function you wrote

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -185,75 +195,483 @@ uint32 insert(struct fsobj *o, struct fsobj *root)
+}
+
+// This is not a full implementation of all that is required by posix
+// but it is enough to pass UT and is all the info relevant to this FS
```
How much work would it be to make this compliant? Also, does this write dummy values for irrelevant fields?

base0x10 replied:
stat has a lot of fields that are not relevant to us.  I used the ones that are slightly useful, but using all the fields would just mean filling in garbage data, like user ID, group ID, timestamps for modification, access, and creation.  etc.  the unit tests only use to it test that a file if it exits returns a pointer and if it does not returns null

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -185,75 +195,483 @@ uint32 insert(struct fsobj *o, struct fsobj *root)
+    }
+    filestats->st_size = file->size;
+    filestats->st_blksize = F_PART_DATA_SIZE;
+    filestats->st_blocks = 1 + file->size % F_PART_DATA_SIZE;
```
I would add some parentheses here

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -185,75 +195,483 @@ uint32 insert(struct fsobj *o, struct fsobj *root)
+int32 file_lseek(int32  filedes, int32 offset, uint32 whence)
+{
+    int32 ret = chk_fd(filedes);
+    if (ret) return ret;
```
Explicitly check against success

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -185,75 +195,483 @@ uint32 insert(struct fsobj *o, struct fsobj *root)
+    return OS_FS_SUCCESS;
+}
+
+int32 file_lseek(int32  filedes, int32 offset, uint32 whence)
```
Should whence be an enum type?

base0x10 replied:
that seems like over abstraction.  its just used as a flag in one location to say from where the offset is starting from.

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -185,75 +195,483 @@ uint32 insert(struct fsobj *o, struct fsobj *root)
+
+    assert(o->position.file_offset == target_offset);
+    return target_offset;
+}
```
It feels like you went a little ham with spacing here. Some is good, but you don't need a space before a comment, for example

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -185,75 +195,483 @@ uint32 insert(struct fsobj *o, struct fsobj *root)
+    uint32 dest_desc = file_dest->filedes;
+
+    char temp_buff[file_src->size];
+    file_read(src_desc, temp_buff,file_src->size);
```
Space around commas

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -185,75 +195,483 @@ uint32 insert(struct fsobj *o, struct fsobj *root)
+    /*
+     * yeah, this is sub-optimal
+     * but I don't expect this to be called at all, much less with large files
+     */
```
Could be a little less informal in this comment. There really isn't a much better way to do this.

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -185,75 +195,483 @@ uint32 insert(struct fsobj *o, struct fsobj *root)
+
+    uint32 dest_desc = file_dest->filedes;
+
+    char temp_buff[file_src->size];
```
This works??? Wow. It might be better to have a global buffer, instead of allocating a lot of data on the stack.

base0x10 replied:
If I made a global buffer then I would need to make it as large as the largest allowable file.  So we would need to both have a filesize limit and also waste that much memory.  
I would rather either a) write a buffered copier, b) assert that the size is less than some defined value, maybe a page or two, and otherwise leave it as is so that the first person who tries to adapt this file system to a system that might need to copy large files realizes that it must be adapted, or c) do nothing and let the end user blow the stack

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -185,75 +195,483 @@ uint32 insert(struct fsobj *o, struct fsobj *root)
+
+int32 file_mv(const char *src, const char *dest)
+{
+
```
No space at the top of functions. Fix this throughout.

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -185,75 +195,483 @@ uint32 insert(struct fsobj *o, struct fsobj *root)
+    if ()
+    struct dirent *p = NULL;i
+    return p;
+}
```
What's going on with this function???

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -185,75 +195,483 @@ uint32 insert(struct fsobj *o, struct fsobj *root)
-    for (i=0 ; i < MAX_NUM_FS && fs[i].devname != NULL ; i++) {
-        if (!strcmp(fs[i].devname, devname)) {
-            fs[i].mountpoint = mountpoint;
+    assert(devname);
```
I'd put the assert above the declaration.

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -185,75 +195,483 @@ uint32 insert(struct fsobj *o, struct fsobj *root)
+    assert(devname);
+    for (i=0 ; i < MAX_NUM_FS && filesystem[i].devname != NULL ; i++) {
+        if (!strcmp(filesystem[i].devname, devname)) {
+
```
Don't put a space after an if... It just looks weird.

base0x10 replied:
perhaps I misunderstand your comment but a space between 'if' and '(' is how cos style guide does it although it isnt specifically said it is required.  I have been doing if_( throughout

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -286,23 +700,75 @@ uint32 newfs_init(char *devname, char *volname, uint32 blocksize, uint32 numbloc
+
+}
+
+int32 Filesys_GetPhysDriveName(char * PhysDriveName, char * MountPoint)
```
Inconsistent naming here

Others reviewed (changes requested):
This is a good PR, but I think it needs some fixes before being merged. Also, this PR is too big. In the future you should submit your code more incrementally. I think Gabe likes PR's of ~400 lines max, and this is 2.5x that. 

In terms of general changes, I think there is some style inconsistency which I've noted in the comments. There are also now 3 layers of API for filesystem stuff, which seems extreme. Can't it just be OS_* stuff, and then internal filesystem stuff? You know you're in trouble when a lot of methods are just wrappers around other methods... My main recommendation here is to reduce the number of weird names. Every non utility function should start with `OS_`, `file_` or `filesys_`. 

Additionally I think you should get rid of all the refcount stuff for now. Concurrent access is going to break things, and having half a system for dealing with it feels useless.

I usually do two passes over a PR before submitting a review, but due to the large size of this one I've only done one. After you fix these ~80 comments, I'll go over it again.

base0x10 reviewed (commented):
Replies to @others review.

gparmer commented:
I'll give more fine-grained comments over the coming days, but at a high-level there are some issues that could have been prevented if we talked about the design throughout the process.  We did not, so you've created a lot of work for yourself ;-)

1. Since we know we want to split the fs stuff off into a library, it seems wise to me to preemptively take this into account.  You already have a wrapper layer that in some sense translates between CFE functions to internal functions, so why not also preemptively save yourself time in the transition that will follow?  Examples include: using the types of the target system in teh correct locations (with casts in-between in the wrappers), using return values appropriate for the layer of abstraction and translating between them in the wrappers.  This is, of course, not necessary, but will lead to a lot of work later.  It puts this PR into a strange place of an intermediate state which makes it hard to focus the review.  This is fine, and I'm mainly pointing out the sw engineering implications of the current design.

2. Related to 1., since you know the fs stuff is going into a cos library, it seems prudent to not deviate too far from the CSG.  The issues of style are (a) using inline declaration of variables everywhere, (b) avoiding using blank lines to delineate the "paragraphs" of your code, (c) basic CSG stuff like conditions/loops, etc...., (d) consistency of type names and function prefixes, (e) comment style, etc...

3. If this code doesn't work (some of the FS tests started failing, right?), then please delineate what works and what doesn't.

4. This is using an old version of the PR template.  The new version asks you to explicitly list out what you used to test your code.  This would likely be useful information to have for the PR and you could spell out 3. explicitly.

base0x10 committed:
fix filesystem directory api and style issues

Implemented many changes suggested during the pull request process.
Many of the changes are style or low impact changes.
The largest change is the directory API.  The directory stream
now behaves as expected, and a few methods have been moved so that
their implementations are all in one place.

zacharied commented on src/components/implementation/no_interface/cFE_booter/osfiles.c:
```diff
@@ -98,120 +106,169 @@ int32 OS_remove(const char *path)
+    char *dest_path = (char *)dest;
+
+    if (!src_path || !dest_path) return OS_FS_ERR_INVALID_POINTER;
+    if (strlen(src_path) > OS_MAX_PATH_LEN) return OS_FS_ERR_PATH_TOO_LONG;
```
Consider compressing these into one `if`

gparmer commented on src/components/implementation/no_interface/cFE_booter/osfiles.c:
```diff
@@ -26,70 +33,71 @@ int32 OS_creat(const char *path, int32  access)
-    if (!path) return OS_FS_ERR_INVALID_POINTER;
-    if (strlen(path) > OS_MAX_PATH_LEN) return OS_FS_ERR_PATH_TOO_LONG;
-    return open(path);
+    int32 ret = chk_path(path);
```
I'm pretty sure that even cFE doesn't want `verb_noun` variable/function names.  Please read the CSG.  This function should at least be `path_chk`, but more likely `fs_path_chk`.

gparmer commented on src/components/implementation/no_interface/cFE_booter/osfiles.c:
```diff
@@ -26,70 +33,71 @@ int32 OS_creat(const char *path, int32  access)
-    PANIC("Read Only Filesystem!");
+    int32 ret = chk_path(path);
+    if (ret != OS_FS_SUCCESS) return ret;
+    file_remove((char *)path);
```
`file_remove` doesn't return an error if the `path` doesn't exist?

gparmer commented on src/components/implementation/no_interface/cFE_booter/osfiles.c:
```diff
@@ -98,120 +106,169 @@ int32 OS_remove(const char *path)
+    if (!fd_prop)                                   return OS_FS_ERR_INVALID_POINTER;
+    if (filedes <= 0 || filedes > MAX_NUM_FILES)    return OS_FS_ERR_INVALID_FD;
+
+    return file_FDGetInfo(filedes, fd_prop);
```
lot of namespaces flying around.  `chk`, `file`, `newfs`, etc...  Not sure if there is a reason to have all of them.  Likely most of them are just `fs` or `mfs` (memory fs).

gparmer commented on src/components/implementation/no_interface/cFE_booter/osfiles.c:
```diff
@@ -98,120 +106,169 @@ int32 OS_remove(const char *path)
+
+    struct fsobj *file = file_find(Filename);
+    if (!file) return OS_INVALID_POINTER;
+    if (!file->refcnt) return OS_FS_ERROR;
```
if `refcnt == 0`, then you should have have access to an allocated file.  So this check either is superfluous, or shows a deeper problem about how you're using refcnts.

gparmer commented on src/components/implementation/no_interface/cFE_booter/osfiles.c:
```diff
@@ -98,120 +106,169 @@ int32 OS_remove(const char *path)
-    PANIC("Unimplemented method!"); // TODO: Implement me!
-    return 0;
+    uint32 i;
+    for (i = 1 ; i < MAX_NUM_FILES + 1 ; i++) {
```
Will look through more code, but starting the iteration at `1` seems strange.

gparmer reviewed (commented):
Yeah, just remembered that you closed this.  No more comments here then!
